etna/lib/etna/clients/magma/workflows/create_project_from_json_workflow.rb

```diff
@@ -0,0 +1,159 @@
+# Creates a project from a JSON file.
+# This workflow:
+# 1) Creates the project in Janus.
+# 2) Adds the user as a project administrator to Janus.
+# 3) Refreshes the user's token with the new privileges.
+# 4) Creates the project in Magma.
+# 5) Iterates through the models and attributes and creates those in Magma.
+
+require 'json'
+require 'ostruct'
+require_relative './model_synchronization_workflow'
+require_relative '../../janus/models'
+require_relative './json_models'
+
+module Etna
+  module Clients
+    class Magma
+      # Note!  These workflows are not perfectly atomic, nor perfectly synchronized due to nature of the backend.
+      # These primitives are best effort locally synchronized, but cannot defend the backend or simultaneous
+      # system updates.
+      class CreateProjectFromJsonWorkflow < Struct.new(:magma_client, :janus_client, :filepath, keyword_init: true)
+        attr_reader :project, :project_name
+
+        def initialize(**params)
+          super({}.update(params))
+          @project = Etna::Clients::Magma::JsonProject.new(filepath: filepath)
+          raise "Project JSON has errors: ", project.errors unless project.valid?
+
+          @project_name = project.name
+        end
+
+        def create_janus_project
+          janus_client.add_project(Etna::Clients::Janus::AddProjectRequest.new(
+            project_name: project_name,
+            project_name_full: project.name_full
+          ))
+        end
+
+        def add_janus_user
+          janus_client.add_user(Etna::Clients::Janus::AddUserRequest.new(
+            project_name: project_name,
+            email: user.email,
+            name: "#{user.first} #{user.last}",
+            role: 'editor'
+          ))
+        end
+
+        def update_janus_permissions
+          janus_client.update_permission(Etna::Clients::Janus::UpdatePermissionRequest.new(
+            project_name: project_name,
+            email: user.email,
+            role: 'administrator'
+          ))
+        end
+
+        def refreshed_token
+          janus_client.refresh_token(Etna::Clients::Janus::RefreshTokenRequest.new).token
+        end
+
+        def update_magma_client_token
+          self.magma_client = Etna::Clients::Magma.new(
+            host: self.magma_client.host, token: refreshed_token)
+        end
+
+        def create_magma_project
+          magma_client.update_model(Etna::Clients::Magma::UpdateModelRequest.new(
+            project_name: project_name,
+            actions: [Etna::Clients::Magma::AddProjectAction.new]))
+        end
+
+        def create_magma_models
+          # Technically ensure_magma_tree should create models as needed,
```

Comment:
When just running `ensure_magma_tree` without creating models beforehand, I ran into issues with `/update_model` where the Magma service would alternate between exceptions about not finding the project or not being able to retrieve models on the project, even though the Magma service had restarted and the project existed in the database. I suspect maybe due to the Sequel cache? However, by creating the models explicitly first, everything worked as expected.